helix-core/src/main/java/org/apache/helix/model/CloudConfig.java

```diff
@@ -0,0 +1,287 @@
+  }
+
+  /**
+   * Get the CLOUD_INFO_PARSER_CLASS_NAME field.
+   * @return CLOUD_INFO_PARSER_CLASS_NAME field.
+   */
+  public String getCloudInfoParserClassName() {
+    return _record.getSimpleField(CloudConfigProperty.CLOUD_INFO_PARSER_CLASS_NAME.name());
+  }
+
+  public static class Builder {
+    private String _clusterName = null;
+    private boolean _cloudEnabled = DEFAULT_CLOUD_ENABLED;
+    private String _cloudID;
+    private List<String> _cloudInfoURLList;
+    private String _cloudInfoFetcherClassName;
+    private String _cloudInfoParserClassName;
+
+    public CloudConfig build() {
+      validate();
+      return new CloudConfig(_clusterName, _cloudEnabled, _cloudID, _cloudInfoURLList,
+          _cloudInfoFetcherClassName, _cloudInfoParserClassName);
+    }
+
+    /**
+     * Default constructor
+     */
+    public Builder() {
+    }
+
+    /**
+     * Constructor with Cluster Name as input
+     * @param clusterName
+     */
+    public Builder(String clusterName) {
+      _clusterName = clusterName;
+    }
+
+    /**
+     * Constructor with CloudConfig as input
+     * @param cloudConfig
+     */
+    public Builder(CloudConfig cloudConfig) {
+      _cloudEnabled = cloudConfig.isCloudEnabled();
+      _cloudID = cloudConfig.getCloudID();
+      _cloudInfoURLList = cloudConfig.getCloudInfoURLs();
+      _cloudInfoFetcherClassName = cloudConfig.getCloudInfoFetcherClassName();
+      _cloudInfoParserClassName = cloudConfig.getCloudInfoParserClassName();
+    }
+
+    public Builder setClusterName(String v) {
+      _clusterName = v;
+      return this;
+    }
+
+    public Builder setCloudEnabled (boolean isEnabled) {
+      _cloudEnabled = isEnabled;
+      return this;
+    }
+
+    public Builder setCloudID (String v) {
+      _cloudID = v;
+      return this;
+    }
+
+    public Builder setCloudInfoURL (List<String> v) {
+      _cloudInfoURLList = v;
+      return this;
+    }
+
+    public Builder addCloudInfoURL (String v) {
+      if (_cloudInfoURLList == null) {
+        _cloudInfoURLList = new ArrayList<String>();
+      }
+      _cloudInfoURLList.add(v);
+      return this;
+    }
+
+    public Builder setCloudInfoFetcherClassName(String v) {
+      _cloudInfoFetcherClassName = v;
+      return this;
+    }
+
+    public Builder setCloudInfoParserClassName(String v) {
+      _cloudInfoParserClassName = v;
+      return this;
+    }
+
+    public String getClusterName() {
+      return _clusterName;
+    }
+
+    public boolean getCloudEnabled() {
+      return _cloudEnabled;
+    }
+
+    public String getCloudID() {
+      return _cloudID;
+    }
+
+    public List<String> getCloudInfoURLList() {
+      return _cloudInfoURLList;
+    }
+
+    public String getCloudInfoFetcherClassName() {
+      return _cloudInfoFetcherClassName;
+    }
+
+    public String getCloudInfoParserClassName() {
+      return _cloudInfoParserClassName;
+    }
+
+    private void validate() {
+      if (_cloudEnabled) {
```

Comment:
So here, users could create invalid CloudConfig objects/ZNodes at will as long as they set `_cloudEnabled` to false. I don't know if that makes too much sense. Is that intended?

Reply:
Yes. It is intended. This feature is enabled once CLOUD_ENABLED is true. If it is false, the feature is disabled.

Reply:
I am not referring to whether a feature is enabled or not; rather, I'm pointing to whether CloudConfig object created is valid or not. These two are separate things. Shouldn't the validation logic validate regardless of whether the feature is enabled or not?

Reply:
You want to have a check for all of the fields? If other fields are null or not? As I see in other builders such as workflow config we do not check all the fields. Do we?
@dasahcc what do you think?

Reply:
User should have the ability to create an empty CLOUDCONFIG before updating something. So if they are doing so, then it means they give up advantage of Helix validation.